src/NServiceBus.Core.Tests/Unicast/MessageOperationsTests.cs

```diff
@@ -0,0 +1,131 @@
+        {
+            var sendPipeline = new FakePipeline<IOutgoingSendContext>();
+            var context = CreateContext(sendPipeline);
+
+            MessageOperations.Send<MyMessage>(context, m => { }, new SendOptions());
+
+            Assert.That(sendPipeline.ReceivedContext.Message.MessageType, Is.EqualTo(typeof(MyMessage)));
+        }
+
+        [Test]
+        public void When_replying_message_interface_should_set_interface_as_message_type()
+        {
+            var replyPipeline = new FakePipeline<IOutgoingReplyContext>();
+            var context = CreateContext(replyPipeline);
+
+            MessageOperations.Reply<IMyMessage>(context, m => { }, new ReplyOptions());
+
+            Assert.That(replyPipeline.ReceivedContext.Message.MessageType, Is.EqualTo(typeof(IMyMessage)));
+        }
+
+        [Test]
+        public void When_replying_message_class_should_set_class_as_message_type()
+        {
+            var replyPipeline = new FakePipeline<IOutgoingReplyContext>();
+            var context = CreateContext(replyPipeline);
+
+            MessageOperations.Reply<MyMessage>(context, m => { }, new ReplyOptions());
+
+            Assert.That(replyPipeline.ReceivedContext.Message.MessageType, Is.EqualTo(typeof(MyMessage)));
+        }
+
+        [Test]
+        public void When_publishing_event_interface_should_set_interface_as_message_type()
+        {
+            var publishPipeline = new FakePipeline<IOutgoingPublishContext>();
+            var context = CreateContext(publishPipeline);
+
+            MessageOperations.Publish<IMyMessage>(context, m => { }, new PublishOptions());
+
+            Assert.That(publishPipeline.ReceivedContext.Message.MessageType, Is.EqualTo(typeof(IMyMessage)));
+        }
+
+        [Test]
+        public void When_publishing_event_class_should_set_class_as_message_type()
+        {
+            var publishPipeline = new FakePipeline<IOutgoingPublishContext>();
+            var context = CreateContext(publishPipeline);
+
+            MessageOperations.Publish<MyMessage>(context, m => { }, new PublishOptions());
+
+            Assert.That(publishPipeline.ReceivedContext.Message.MessageType, Is.EqualTo(typeof(MyMessage)));
+        }
+
+        IBehaviorContext CreateContext<TContext>(IPipeline<TContext> pipeline) where TContext : IBehaviorContext
+        {
+            var pipelineCache = new FakePipelineCache();
+            pipelineCache.RegisterPipeline(pipeline);
+
+            var context = new TestableMessageHandlerContext();
+            context.Builder.Register<IMessageMapper>(() => new MessageMapper());
+            context.Extensions.Set<IPipelineCache>(pipelineCache);
+
+            return context;
+        }
+
+        // ReSharper disable once MemberCanBePrivate.Global
+        public interface IMyMessage
+        {
+        }
+
+        class MyMessage
+        {
+        }
+
+        class FakePipelineCache : IPipelineCache
```

Comment:
I think this is not the only pipeline cache we have. Would it make sense in a separate PR to toolify this?